python/ray/serve/_private/router.py

```diff
@@ -205,45 +240,87 @@ def max_num_scheduling_tasks(self) -> int:
 
     @property
     def target_num_scheduling_tasks(self) -> int:
-        """Target number of scheduling tasks to be running based on pending assignments.
+        """Target number of scheduling tasks to be running based on pending requests.
 
         This will never exceed `self.max_num_scheduling_tasks`.
         """
-        return min(self.num_pending_assignments, self.max_num_scheduling_tasks)
+        return min(self.num_pending_requests, self.max_num_scheduling_tasks)
+
+    @property
+    def curr_replicas(self) -> Dict[str, ReplicaWrapper]:
+        return self._replicas
 
     def update_replicas(self, replicas: List[ReplicaWrapper]):
         """Update the set of available replicas to be considered for scheduling.
 
         When the set of replicas changes, we may spawn additional scheduling tasks
-        if there are pending assignments.
+        if there are pending requests.
         """
-        self._replicas = {r.replica_id: r for r in replicas}
-        new_replica_id_set = set(self._replicas.keys())
+        new_replicas = {}
+        new_replica_id_set = set()
+        new_multiplexed_model_id_to_replica_ids = defaultdict(set)
+        for r in replicas:
+            new_replicas[r.replica_id] = r
+            new_replica_id_set.add(r.replica_id)
+            for model_id in r.multiplexed_model_ids:
+                new_multiplexed_model_id_to_replica_ids[model_id].add(r.replica_id)
+
+        self._replicas = new_replicas
+        self._replica_id_set = new_replica_id_set
+        self._multiplexed_model_id_to_replica_ids = (
+            new_multiplexed_model_id_to_replica_ids
+        )
+        self._replicas_updated_event.set()
+        self.maybe_start_scheduling_tasks()
+
         if self._replica_id_set != new_replica_id_set:
```

Comment:
This statement will always be `False`. It needs to be run before [`self._replica_id_set = new_replica_id_set`](https://github.com/ray-project/ray/pull/36874/files#diff-b81ab953bc243bf2c453db721719903571d53fdd13d9607a292a4658951d66f0R269).

Reply:
ah good catch (this is only for logging, so not caught by tests)